### DISCUSSION TITLE
Which redirect approach to use when?

### DISCUSSION THREAD

**Discussion by TusharShahi** [vercel/next.js#40059] [Help]
There are multiple ways to redirect from one page to another. Say we have a `/home` page where we want our `/` route to redirect to. There are a few options:

1. `getServerSideProps` in the index.js file in pages folder.
2. Using redirect in middleware using NextResponse
3. Using next.config.js redirects

Using the third approach complex logic can't be used. But in simple cases as above, which approach should be used when?

**Comment by icyJoseph** [comment]
Few late night thoughts. Might update later.

> 1. `getServerSideProps` in the index.js file in pages folder.

If you need to look at some data, use Node.js APIs, or if `404` is also a possible result,  use `getServerSideProps`.

Middleware can't really use Node.js APIs fully, and since `12.2.x` it can't respond with body to the caller, so direct 404's are no longer plausible. 

Very easy to test.

You can report to analytics services that `/path` has been redirect to `/other-path`.

> 2. Using redirect in middleware using NextResponse

Again, if you need to look at some data, or draw a conclusion from the incoming request, this is a good place to do it. Although, middleware, depending on the use case, could end up doing longer round trips to find database and other services. 

Relatively easy to test. 

You can also report to analytics services.

> 3. Using next.config.js redirects

If no complex logic is needed, or the redirect rules are known upfront, then the config redirects are good enough, though you miss out on testing them with Jest, and often you might need to write regex matchers, which are hard to reason about.
